File: balanced_binary_tree/balanced_binary_tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
   return dfs(root) != -1
}

func dfs(root *TreeNode) int {
   if root == nil {
       return 0
   }
   leftH := dfs(root.Left)
   rightH := dfs(root.Right)

   if leftH == -1 || rightH == -1 || abs(leftH - rightH) > 1 {
       return -1
   }

   return max(leftH, rightH) + 1
}

func max(a, b int) int {
   if a > b {
       return a
   }
   return b
}

func abs(n int) int {
   if n > 0 {
       return n
   }
   return -n
}

File: balanced_binary_tree/balanced_binary_tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn is_balanced(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        fn dfs(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
            match root {
                Some(root) => {
                    let leftH = dfs(root.borrow().left.clone());
                    let rightH = dfs(root.borrow().right.clone());
                    if leftH == -1 || rightH == -1 || (leftH - rightH).abs() > 1 {
                        return -1
                    }
                    leftH.max(rightH) + 1
                }

                None => 0
            }
        }
        dfs(root) != -1
    }
}